src/mysqlsh_helpers.py

```diff
@@ -0,0 +1,176 @@
+            root_password: password for the 'root' user
+            cluster_admin_user: user name for the cluster admin user
+            cluster_admin_password: password for the cluster admin user
+            instance_address: address of the targeted instance
+        """
+        self.root_password = root_password
+        self.cluster_admin_user = cluster_admin_user
+        self.cluster_admin_password = cluster_admin_password
+        self.instance_address = instance_address
+
+    def mysqlsh_bin(self, container: Container) -> str:
+        """Determine binary path for MySQL Shell.
+
+        Args:
+            container: Container to run the script in
+
+        Returns:
+            Path to binary mysqlsh
+        """
+        # Allow for various versions of the mysql-shell snap
+        # When we get the alias use /snap/bin/mysqlsh
+        _paths = ("/snap/bin/mysqlsh", "/snap/bin/mysql-shell.mysqlsh")
+
+        for path in _paths:
+            if container.exists(path):
+                return path
+        # Default to the upstream repo mysql-shell
+        return "/usr/bin/mysqlsh"
+
+    def configure_mysql_users(self) -> None:
+        """Configure the MySQL users for the instance.
+
+        Creates base `root@%` and `clusteradmin@instance_address` user with the
+        appropriate privileges, and reconfigure `root@localhost` user password.
+        Raises MySQLConfigureMySQLUsersError if the user creation fails.
+        """
+        _script = (
+            "SET @@SESSION.SQL_LOG_BIN=0;",
+            f"CREATE USER '{self.cluster_admin_user}'@'{self.instance_address}' IDENTIFIED BY '{self.cluster_admin_password}';",
+            f"GRANT ALL ON *.* TO '{self.cluster_admin_user}'@'{self.instance_address}' WITH GRANT OPTION;",
+            f"CREATE USER 'root'@'%' IDENTIFIED BY '{self.root_password}';",
+            "GRANT ALL ON *.* TO 'root'@'%' WITH GRANT OPTION;",
+            "UPDATE mysql.user SET authentication_string=null WHERE User='root' and Host='localhost';",
+            f"ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY '{self.root_password}';",
+            "REVOKE SYSTEM_USER ON *.* FROM root@'%';",
+            "REVOKE SYSTEM_USER ON *.* FROM root@localhost;",
+            "FLUSH PRIVILEGES;",
+        )
+
+        try:
+            self._run_mysqlcli_script(" ".join(_script))
+        except ExecError as e:
+            logger.exception(f"Failed to configure users for: {self.instance_address}", exc_info=e)
+            raise MySQLConfigureMySQLUsersError(e.stdout)
+
+    def configure_instance(self) -> None:
+        """Configure the instance to be used in an InnoDB cluster."""
+        commands = (
+            f"dba.configure_instance('{self.cluster_admin_user}:{self.cluster_admin_password}@{self.instance_address}')",
+            f"my_shell = shell.connect('{self.cluster_admin_user}:{self.cluster_admin_password}@{self.instance_address}')",
+            'my_shell.run_sql("RESTART;");',
```

Comment:
i ran into an issue that `RESTART` tries to restart with systemd. we'd need to confirm this is the case, and if so, we'd need to restart with pebble

Reply:
That's a miss, well remembered.